File: java/src/main/java/benchmarks/GameOfLife.java
package benchmarks;

import java.util.*;

public class GameOfLife extends Benchmark {
    private static class Cell {
        private boolean alive;
        private boolean nextState;
        private Cell[] neighbors;
        private int neighborCount;

        public Cell(boolean alive) {
            this.alive = alive;
            this.nextState = false;
            this.neighbors = new Cell[8];
        }

        public void addNeighbor(Cell cell) {
            neighbors[neighborCount++] = cell;
        }

        public void computeNextState() {
            int aliveNeighbors = 0;
            for (Cell neighbor : neighbors) {
                if (neighbor.alive) aliveNeighbors++;
            }

            if (alive) {
                nextState = (aliveNeighbors == 2 || aliveNeighbors == 3);
            } else {
                nextState = (aliveNeighbors == 3);
            }
        }

        public void update() {
            alive = nextState;
        }

        public void setAlive(boolean state) {
            alive = state;
        }

        public boolean isAlive() {
            return alive;
        }
    }

    private static class Grid {
        private final int width;
        private final int height;
        private final Cell[][] cells;

        public Grid(int width, int height) {
            this.width = width;
            this.height = height;
            this.cells = new Cell[height][width];

            for (int y = 0; y < height; y++) {
                for (int x = 0; x < width; x++) {
                    cells[y][x] = new Cell(false);
                }
            }

            linkNeighbors();
        }

        private void linkNeighbors() {
            for (int y = 0; y < height; y++) {
                for (int x = 0; x < width; x++) {
                    Cell cell = cells[y][x];

                    for (int dy = -1; dy <= 1; dy++) {
                        for (int dx = -1; dx <= 1; dx++) {
                            if (dx == 0 && dy == 0) continue;

                            int ny = (y + dy + height) % height;
                            int nx = (x + dx + width) % width;

                            cell.addNeighbor(cells[ny][nx]);
                        }
                    }
                }
            }
        }

        public void nextGeneration() {
            for (Cell[] row : cells) {
                for (Cell cell : row) {
                    cell.computeNextState();
                }
            }

            for (Cell[] row : cells) {
                for (Cell cell : row) {
                    cell.update();
                }
            }
        }

        public int countAlive() {
            int count = 0;
            for (int y = 0; y < height; y++) {
                for (int x = 0; x < width; x++) {
                    if (cells[y][x].isAlive()) count++;
                }
            }
            return count;
        }

        public long computeHash() {
            final long FNV_OFFSET_BASIS = 2166136261L;
            final long FNV_PRIME = 16777619L;

            long hash = FNV_OFFSET_BASIS;
            for (int y = 0; y < height; y++) {
                for (int x = 0; x < width; x++) {
                    long alive = cells[y][x].isAlive() ? 1L : 0L;
                    hash = (hash ^ alive) * FNV_PRIME;
                }
            }
            return hash;
        }

        public Cell[][] getCells() {
            return cells;
        }
    }

    private final int width;
    private final int height;
    private Grid grid;

    public GameOfLife() {
        this.width = (int) configVal("w");
        this.height = (int) configVal("h");
        this.grid = new Grid(width, height);
    }

    @Override
    public String name() {
        return "Etc::GameOfLife";
    }

    @Override
    public void prepare() {
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                if (Helper.nextFloat() < 0.1f) {
                    grid.getCells()[y][x].setAlive(true);
                }
            }
        }
    }

    @Override
    public void run(int iterationId) {
        grid.nextGeneration();
    }

    @Override
    public long checksum() {
        int alive = grid.countAlive();
        return grid.computeHash() + alive;
    }
}

File: scala/src/main/scala/benchmarks/GameOfLife.scala
package benchmarks

class GameOfLife extends Benchmark:
  private class CellObj:
    var alive: Boolean = false
    var nextState: Boolean = false
    val neighbors = new Array[CellObj](8)
    var neighborCount: Int = 0

    def addNeighbor(cell: CellObj): Unit =
      neighbors(neighborCount) = cell
      neighborCount += 1

    def computeNextState(): Unit =
      var aliveNeighbors = 0
      var i = 0
      while i < neighborCount do
        if neighbors(i).alive then aliveNeighbors += 1
        i += 1

      nextState =
        if alive then aliveNeighbors == 2 || aliveNeighbors == 3
        else aliveNeighbors == 3

    def update(): Unit =
      alive = nextState

  private class Grid(val width: Int, val height: Int):
    private val cells = Array.ofDim[CellObj](height, width)

    for
      y <- 0 until height
      x <- 0 until width
    do cells(y)(x) = new CellObj

    for
      y <- 0 until height
      x <- 0 until width
    do
      val cell = cells(y)(x)
      for
        dy <- -1 to 1
        dx <- -1 to 1
        if !(dx == 0 && dy == 0)
      do
        val ny = (y + dy + height) % height
        val nx = (x + dx + width) % width
        cell.addNeighbor(cells(ny)(nx))

    def nextGeneration(): Unit =

      for
        y <- 0 until height
        x <- 0 until width
      do cells(y)(x).computeNextState()

      for
        y <- 0 until height
        x <- 0 until width
      do cells(y)(x).update()

    def countAlive(): Int =
      var count = 0
      for
        y <- 0 until height
        x <- 0 until width
      do if cells(y)(x).alive then count += 1
      count

    def computeHash(): Long =
      val FNV_OFFSET_BASIS = 2166136261L
      val FNV_PRIME = 16777619L

      var hasher = FNV_OFFSET_BASIS
      for
        y <- 0 until height
        x <- 0 until width
      do
        val alive = if cells(y)(x).alive then 1L else 0L
        hasher = (hasher ^ alive) * FNV_PRIME

      hasher & 0xffffffffL

    def getCells(): Array[Array[CellObj]] = cells

  private val width: Int = configVal("w").toInt
  private val height: Int = configVal("h").toInt
  private var grid: Grid = _

  override def name(): String = "Etc::GameOfLife"

  override def prepare(): Unit =
    grid = Grid(width, height)

    for
      y <- 0 until height
      x <- 0 until width
    do if Helper.nextFloat() < 0.1f then grid.getCells()(y)(x).alive = true

  override def run(iterationId: Int): Unit =
    grid.nextGeneration()

  override def checksum(): Long =
    val alive = grid.countAlive()
    (grid.computeHash() + alive) & 0xffffffffL
